code/trans.c

```diff
@@ -0,0 +1,377 @@
+  AVERT_CRITICAL(ScanState, ss);
+  AVER_CRITICAL(refIO != NULL);
+
+  transform = ss->fixClosure;
+  AVERT_CRITICAL(Transform, transform);
+
+  if (!transform->aborted) {
+    void *refNew;
+
+    ref = *refIO;
+
+    if (TableLookup(&refNew, transform->oldToNew, (Word)ref)) {
+      if (ss->rank == RankAMBIG) {
+        /* We rely on the fact that ambiguous references are fixed
+           first, so that no exact references have been transformed
+           yet. */
+        transform->aborted = TRUE;
+      } else {
+        /* NOTE: We could fix refNew in the table before copying it,
+           since any summaries etc. collected in the scan state will still
+           apply when it's copied.  That could save a few snap-outs. */
+        *refIO = refNew;
+      }
+    }
+  }
+
+  /* Now progress to a normal GC fix. */
+  /* TODO: Make a clean interface to this kind of dynamic binding. */
+  ss->fix = ss->arena->emergency ? SegFixEmergency : SegFix;
+  TRACE_SCAN_BEGIN(ss) {
+    res = TRACE_FIX12(ss, refIO);
+  } TRACE_SCAN_END(ss);
+  ss->fix = transformFix;
+
+  return res;
+}
+
+
+static void transformCondemn(void *closure, Word old, void *value)
+{
+  Seg seg;
+  GenDesc gen;
+  Bool b;
+  Trace trace = closure;
+
+  AVERT(Trace, trace);
+  UNUSED(value);
+
+  /* Find segment containing old address. */
+  b = SegOfAddr(&seg, trace->arena, (Ref)old);
+  AVER(b); /* old refs must be in managed memory, else client param error */
+
+  /* Condemn generation containing seg if not already condemned. */
+  gen = PoolSegPoolGen(SegPool(seg), seg)->gen;
+  AVERT(GenDesc, gen);
+  if (RingIsSingle(&gen->trace[trace->ti].traceRing))
+    GenDescStartTrace(gen, trace);
+}
+
+
+Res TransformApply(Bool *appliedReturn, Transform transform)
+{
+  Res res;
+  Arena arena;
+  Globals globals;
+  Trace trace;
+  double mortality;
+
+  AVER(appliedReturn != NULL);
+  AVERT(Transform, transform);
+
+  arena = TransformArena(transform);
+
+  /* If there have been any flips since the transform was created, the old
+     and new pointers will be invalid, since they are not scanned as roots.
+     The client program must park the arena before applying the transform. */
+  if (transform->epoch != ArenaEpoch(arena))
+    return ResPARAM;
+
+  globals = ArenaGlobals(arena);
+  AVERT(Globals, globals);
+
+  ArenaPark(globals);
+  
+  res = TraceCreate(&trace, arena, TraceStartWhyEXTENSION);
+  AVER(res == ResOK); /* parking should make a trace available */
+  if (res != ResOK)
+    return res;
+
+  /* Condemn the generations containing the transform's old objects,
+     so that all references to them are scanned. */
+  TraceCondemnStart(trace);
+  TableMap(transform->oldToNew, transformCondemn, trace);
+  res = TraceCondemnEnd(&mortality, trace);
+  if (res != ResOK) {
+    /* Nothing to transform. */
+    TraceDestroyInit(trace);
+    goto done;
+  }
+
+  trace->fix = transformFix;
+  trace->fixClosure = transform;
+
+  res = TraceStart(trace, 1.0, 0.0);
```

Comment:
These estimates of the mortality and finishing time do not seem very good! Could we compute better ones in `transformCondemn`?